### DISCUSSION TITLE
QST: memory consumption with ArrowDtype structs

### DISCUSSION THREAD

**Discussion by hilaaba** [geopandas/geopandas#3516] [Q&A] [answered]
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of geopandas.

- [ ] (optional) I have confirmed this bug exists on the main branch of geopandas.

---

**Context**:

Our company stores geospatial data in MongoDB. Recently, we have been exploring the use of [pymongoarrow](https://www.mongodb.com/docs/languages/python/pymongo-arrow-driver/current/) for loading data with [find_arrow_all()](https://mongo-arrow.readthedocs.io/en/1.6.0/api/api.html#pymongoarrow.api.find_arrow_all). Our research has shown that `pymongoarrow` efficiently loads data. Additionally, we aim to ensure that the columns use [pd.ArrowDtype](https://pandas.pydata.org/docs/reference/api/pandas.ArrowDtype.html) for efficient in-memory representation.

**Use case**:

```python
table = find_arrow_all(...)
df = table.to_pandas(types_mapper=pd.ArrowDtype)
df["gpd_geometry"] = df["geometry"].apply(shape)
gdf = gpd.GeoDataFrame(df, geometry="gpd_geometry", crs="epsg:4326")
```
However, when performing spatial joins on the resulting GeoDataFrame, we observe unusually high memory usage with ArrowDtype.

To demonstrate the issue, we perform a spatial join between random points and polygons located within the same region.

```python
import math
import random

import geopandas as gpd
import pandas as pd
import pyarrow as pa
from shapely.geometry import shape


def generate_random_points(count, lon_range, lat_range):
    points = []
    for num in range(count):
        points.append(
            {
                'id': num,
                'geometry': {
                    'type': 'Point',
                    'coordinates': [
                        random.uniform(*lon_range),
                        random.uniform(*lat_range),
                    ],
                },
            }
        )
    return points


def generate_random_polygons(count, lon_range, lat_range, radius_range, vertices_count_range):
    polygons = []
    for num in range(count):
        vertices_count = random.randint(*vertices_count_range)

        center_lon = random.uniform(*lon_range)
        center_lat = random.uniform(*lat_range)

        angles = sorted([random.uniform(0, 2 * math.pi) for _ in range(vertices_count)])
        radius = [random.uniform(*radius_range) for _ in range(vertices_count)]

        vertices = [
            [center_lon + r * math.cos(angle), center_lat + r * math.sin(angle)]
            for r, angle in zip(radius, angles)
        ]
        vertices.append(vertices[0])

        polygons.append(
            {
                'id': num,
                'geometry': {
                    'type': 'Polygon',
                    'coordinates': [vertices],
                },
            }
        )
    return polygons


def get_gdf(items, geom_name='gpd_geometry', use_pyarrow_types=True):
    table = pa.Table.from_pylist(items)
    df = table.to_pandas(types_mapper=pd.ArrowDtype if use_pyarrow_types else None)
    df[geom_name] = df['geometry'].apply(shape)
    return gpd.GeoDataFrame(df, geometry=geom_name, crs='epsg:4326')


lon_range = (5, 15)
lat_range = (45, 55)

points = generate_random_points(count=100_000, lon_range=lon_range, lat_range=lat_range)
polygons = generate_random_polygons(
    count=10_000,
    lon_range=lon_range,
    lat_range=lat_range,
    radius_range=(0.5, 1.0),
    vertices_count_range=(50, 100),
)

points_gdf = get_gdf(points)
polygons_gdf = get_gdf(polygons)
result = gpd.sjoin(
    points_gdf,
    polygons_gdf,
    how='left',
    predicate='intersects',
)
```

#### Problem description

When executing the spatial join with the above code, we notice that memory usage spikes significantly. This issue is observed specifically when the columns are typed using pd.ArrowDtype. Here is the RAM and CPU graph:
![fail_case](https://github.com/user-attachments/assets/4e34084f-2119-4b72-ab49-dcabe9dd21fa)

Interestingly, the issue disappears if we use the default column name geometry when creating the GeoDataFrame. For example, modifying the `geom_name` parameter as follows resolves the problem:

```python
points_gdf = get_gdf(points, geom_name='geometry')
polygons_gdf = get_gdf(polygons, geom_name='geometry')
result = gpd.sjoin(
    points_gdf,
    polygons_gdf,
    how='left',
    predicate='intersects',
)
```

Another observation is that the issue also disappears when we avoid using pd.ArrowDtype. For example:

```python
points_gdf = get_gdf(points, use_pyarrow_types=False)
polygons_gdf = get_gdf(polygons, use_pyarrow_types=False)
result = gpd.sjoin(
    points_gdf,
    polygons_gdf,
    how='left',
    predicate='intersects',
)
```

The graph of these two cases will look like this:
![success_case](https://github.com/user-attachments/assets/509c6a01-34c9-48bc-8d78-5fc87221cda2)

#### Expected Output

In my opinion, the `sjoin` method should work correctly regardless of the column types or the column name for geometry.

<details>

SYSTEM INFO
-----------
python     : 3.11.6 | packaged by conda-forge | (main, Oct  3 2023, 11:57:02) [GCC 12.3.0]
executable : /opt/conda/bin/python
machine    : Linux-6.6.32-linuxkit-aarch64-with-glibc2.35

GEOS, GDAL, PROJ INFO
---------------------
GEOS       : 3.11.4
GEOS lib   : None
GDAL       : 3.9.1
GDAL data dir: /opt/conda/lib/python3.11/site-packages/pyogrio/gdal_data/
PROJ       : 9.4.1
PROJ data dir: /opt/conda/lib/python3.11/site-packages/pyproj/proj_dir/share/proj

PYTHON DEPENDENCIES
-------------------
geopandas  : 1.0.1
numpy      : 2.1.3
pandas     : 2.2.3
pyproj     : 3.7.0
shapely    : 2.0.6
pyogrio    : 0.10.0
geoalchemy2: None
geopy      : None
matplotlib : 3.9.2
mapclassify: None
fiona      : None
psycopg    : None
psycopg2   : None
pyarrow    : 18.0.0

</details>

**Comment by m-richards** [comment]
Hi @hilaaba, this was a bit of a surprising issue to look at, and is actually a lot simpler than it first seemed. The memory issues you see are likely a real issue, but most likely in pyarrow.

These are some outputs from modifying your example a bit:
```python
points_gdf = get_gdf(points, geom_name="gpd_geometry", use_pyarrow_types=True)
print(points_gdf.dtypes)
id                                                 int64[pyarrow]
geometry        struct<coordinates: list<item: double>, type: ...
gpd_geometry                                             geometry
dtype: object
points_gdf = get_gdf(points, geom_name="geometry", use_pyarrow_types=True)
print(points_gdf.dtypes)
id          int64[pyarrow]
geometry          geometry
dtype: object
points_gdf = get_gdf(points, geom_name="geometry", use_pyarrow_types=False)
print(points_gdf.dtypes)
id             int64
geometry    geometry
dtype: object
```

The reason you see different performance characteristics based on the geometry column name is not to do with the name itself, it's to do with whether the pyarrow table struct column named "geometry" is present in the table. When you give `get_gdf(..., geom_name="geometry")` you overwrite that column, and so the memory issue disappears.

Now, for the actual memory issue itself, this happens after we've done the spatial indexing part of the join, but in the internal `_frame_join` function in `sjoin.py` we call `right = right_df._reindex_with_indexers({0: (new_index, r_idx)})`. Following the call chain down, it's calling 

```python
import pandas.core.algorithms as algos
new_values = algos.take_nd(    values, indexer, axis=axis, allow_fill=allow_fill, fill_value=fill_value)`
```
inside `pandas.core.internals.blocks::Block::take_nd` that seems to use a lot more memory than the object dtype case.

Unfortunately, this isn't exactly a nice minimal reproducer to go reporting upstream, I'll see if I can come back to this and extract a small reproducer.

**Comment by m-richards** [comment]
Here's an equivalent at least without any geopandas/ shapely / random numbers:
```python
import pandas as pd
import pyarrow as pa

def get_df(count, use_pyarrow_types=True):
    structs = []
    for num in range(count):
        structs.append(
            {
                'id': num,
                'geometry': {
                    'type': 'Polygon',
                    'coordinates': [(i, i - 1) for i in range(50)],
                },
            }
        )

    table = pa.Table.from_pylist(structs)
    df = table.to_pandas(types_mapper=pd.ArrowDtype if use_pyarrow_types else None)
    return df

df = get_df(10_000)

df2 = pd.DataFrame({"id":list(i for i in range(1,100_000, 3))})
r_idx = pd.Series(pd.RangeIndex(0, 15_000_000)) %4

right = df._reindex_with_indexers({0: (r_idx.index, r_idx.values)})

```
Edit: swapping the last line for:
`right = df.reindex(r_idx)` also does the same thing but is expressed in terms of the public api.

Also, if 'coordinates' is a list of integers, rather than a list of list of integers the memory performance seems quite a lot better.

**Comment by jorisvandenbossche** [comment, marked as answer]
Thanks for diving in!

> The reason you see different performance characteristics based on the geometry column name is not to do with the name itself, it's to do with whether the pyarrow table struct column named "geometry" is present in the table. When you give `get_gdf(..., geom_name="geometry")` you overwrite that column, and so the memory issue disappears.
> ...
> inside `pandas.core.internals.blocks::Block::take_nd` that seems to use a lot more memory than the object dtype case.

Indeed, it turns out that the `take` operation (i.e. "taking" or selecting certain rows based on indices, which is used under the hood when joining dataframes) is quite expensive for the pyarrow struct array (both in time and memory), at least compared to having the same data in an object dtype object.

The explanation for this is that in case of the struct array, an actual new array with new data gets constructed, while in case of `object` dtype, which stores references to Python objects, we essentially just construct a new array referencing the same Python objects.

In the smaller reproducer above, this becomes a quite extreme difference, because it takes repeatedly the first 4 elements of the original array. With the pyarrow struct array, the result is a fully materialized array that takes 14GB in memory, while for the object array, it just points to 4 python dict repeatedly, and so hardly uses much memory (for the final result, but so also not during the operation). See the example below for some numbers.

---

A further simplified reproducer using the same data but with just pyarrow's `take`:

```python
import numpy as np
import pyarrow as pa

def create_struct_array(count):
    structs = []
    for _ in range(count):
        structs.append(
            {
                'type': 'Polygon',
                'coordinates': [(i, i - 1) for i in range(50)],
            }
        )

    return pa.array(structs)

arr = create_struct_array(10_000)
indexer = np.arange(15_000_000) % 4

arr.take(indexer)
```

which gives

```python
In [2]: %time res1 = arr.take(indexer)
CPU times: user 8.3 s, sys: 8.63 s, total: 16.9 s
Wall time: 16.9 s

In [3]: res1.nbytes / 1024**3
Out[3]: 14.27019014954567

In [4]: %time res2 = arr.to_numpy(zero_copy_only=False).take(indexer)
CPU times: user 105 ms, sys: 65.1 ms, total: 171 ms
Wall time: 170 ms

In [5]: res2.nbytes / 1024**3
Out[5]: 0.11175870895385742
```

And running the pyarrow case with memray:
![newplot](https://github.com/user-attachments/assets/fc49d405-8ca7-4c37-8a9b-66f34cb55ebb)

So during the operation it still uses more memory than the final result, but it's also the final result that just takes a lot of memory.

**Comment by jorisvandenbossche** [comment]
In general we use `pyarrow` to speed things up, but of course that always comes with caveats. And this case is a good example of where the usage of `ArrowDtype` is not actually beneficial, or at least where you should avoid keeping the original struct columns after converting them to actual geometries (as I assume you no longer need them).
